crates/primitives/src/forkid.rs

```diff
@@ -267,36 +327,43 @@ pub struct ForkTransition {
 struct Cache {
     // An epoch is a period between forks.
     // When we progress from one fork to the next one we move to the next epoch.
-    epoch_start: BlockNumber,
-    epoch_end: Option<BlockNumber>,
-    past: Vec<(BlockNumber, ForkHash)>,
+    epoch_start: ForkFilterKey,
+    epoch_end: Option<ForkFilterKey>,
+    past: Vec<(ForkFilterKey, ForkHash)>,
     future: Vec<ForkHash>,
     fork_id: ForkId,
 }
 
 impl Cache {
     /// Compute cache.
-    fn compute_cache(forks: &BTreeMap<BlockNumber, ForkHash>, head: BlockNumber) -> Self {
+    fn compute_cache(forks: &BTreeMap<ForkFilterKey, ForkHash>, head: Head) -> Self {
         let mut past = Vec::with_capacity(forks.len());
         let mut future = Vec::with_capacity(forks.len());
 
-        let mut epoch_start = 0;
+        let mut epoch_start = ForkFilterKey::Block(0);
         let mut epoch_end = None;
-        for (block, hash) in forks {
-            if *block <= head {
-                epoch_start = *block;
-                past.push((*block, *hash));
+        for (key, hash) in forks {
+            let active = if let ForkFilterKey::Block(block) = key {
+                *block <= head.number
+            } else if let ForkFilterKey::Time(time) = key {
+                *time <= head.timestamp
+            } else {
+                unreachable!()
+            };
+            if active {
+                epoch_start = *key;
```

Comment:
i see so this may trigger it to actually be a `ForkFilter::Time` if the active key is that